turbopack/crates/turbo-tasks-macros-shared/src/self_filter.rs

```diff
@@ -0,0 +1,63 @@
+use proc_macro2::TokenTree;
+use quote::ToTokens;
+use syn::{
+    visit::{visit_expr, visit_macro, Visit},
+    Block, Expr, Macro,
+};
+
+pub fn is_self_used(block: &Block) -> bool {
+    let mut finder = SelfFinder { found: false };
+    finder.visit_block(block);
+    finder.found
+}
+
+struct SelfFinder {
+    found: bool,
+}
+
+impl Visit<'_> for SelfFinder {
```

Comment:
Skipping `visit_item_impl` like we do in `RewriteSelfVisitMut` should be safe here as well.